modules/auxiliary/gather/query_ldap.rb

```diff
@@ -0,0 +1,188 @@
+          entries = perform_ldap_query(ldap, filter)
+
+        when 'ENUM_DOMAIN_CONTROLERS'
+          filter = Net::LDAP::Filter.construct('(&(objectCategory=Computer)(userAccountControl:1.2.840.113556.1.4.803:=8192))') # Find domain controllers
+          entries = perform_ldap_query(ldap, filter)
+
+        when 'ENUM_EXCHANGE_SERVERS'
+          filter = Net::LDAP::Filter.construct('(&(objectClass=msExchExchangeServer)(!(objectClass=msExchExchangeServerPolicy)))') # Find Exchange Servers
+          entries = perform_ldap_query(ldap, filter)
+
+        when 'ENUM_EXCHANGE_RECIPIENTS'
+          # Find Exchange Recipients with or without fax addresses.
+          filter = Net::LDAP::Filter.construct('(|(mailNickname=*)(proxyAddresses=FAX:*))')
+          entries = perform_ldap_query(ldap, filter)
+
+        when 'ENUM_GROUPS'
+          # Standard LDAP groups query, followed by trying to find AD security groups, then trying to find Linux groups.
+          # Filters combined to remove duplicates.
+          filter = Net::LDAP::Filter.construct('(|(objectClass=group)(objectClass=groupOfNames)(groupType:1.2.840.113556.1.4.803:=2147483648)(objectClass=posixGroup))')
+          entries = perform_ldap_query(ldap, filter)
+
+        when 'ENUM_ORGUNITS'
+          filter = Net::LDAP::Filter.construct('(objectClass=organizationalUnit)') # Find OUs aka Organizational Units
+          entries = perform_ldap_query(ldap, filter)
+
+        when 'ENUM_ORGROLES'
+          filter = Net::LDAP::Filter.construct('(objectClass=organizationalRole)') # Find OUs aka Organizational Units
+          entries = perform_ldap_query(ldap, filter)
+
+        when 'ENUM_PEOPLE'
+          filter = Net::LDAP::Filter.construct('(objectClass=organizationalPerson)') # Find people within an organization by Person entries.
+          entries = perform_ldap_query(ldap, filter)
+
+        when 'ENUM_USERS'
+          # Common LDAP user query, followed by a query for AD User records by account type.
+          # Finally, query for Linux accounts by objectClass.
+          #
+          # Doing this all in one query also prevents duplicate entires across multiple queries.
+          filter = Net::LDAP::Filter.construct('(|(objectClass=inetOrgPerson)(objectClass=user)(sAMAccountType=805306368)(objectClass=posixAccount)(objectClass=GsAccount)(objectClass=GsSIPUser))')
+          entries = perform_ldap_query(ldap, filter)
+        end
+      end
+    rescue Rex::ConnectionTimeout => e
+      print_error("Could not query #{datastore['RHOST']}! Error was: #{e.message}")
+      return
+    end
+
+    return if entries[1].nil?
+
+    columns = []
+    entries[1].each do |entry|
+      entry.attribute_names.each do |attribute|
+        columns << attribute.to_s
+      end
+    end
+    columns.uniq!
+    tbl = Rex::Text::Table.new(
+      'Header' => "#{action.name} Dump of #{peer}",
+      'Indent' => 1,
+      'Columns' => columns
```

Comment:
I noticed while debugging the issue I ran into that we're building this table with 203 columns (at least when targeting my test AD server). I'm not sure how useful it'll be with all of that data, so let's find a way to trim it down to the most useful columns. It would make sense to have that be specific to the query that's being run.

I think it'd be really helpful if users could provide their own data files with queries defined in like JSON or YAML.

Using `ENUM_USERS` as an example, it's definition might look something like:
```yaml
- name: ENUM_USERS
  filter: (|(objectClass=inetOrgPerson)(objectClass=user)(sAMAccountType=805306368)(objectClass=posixAccount)(objectClass=GsAccount)(objectClass=GsSIPUser))
  columns: 
    - dn
    - objectclass
    - objectsid
```

This would help users distribute their own queries and reuse them.

Reply:
Different LDAP servers may have different columns for the data so this is why I was initially against this idea since I thought we were making a more generic solution. In theory we could do this but then one runs the risk of missing out on potentially valuable information merely cause they forgot another column could be important. 

I'm open to doing filtering this way though but it would need to be planned out and take into account potential server differences.

Reply:
Active Directory should be pretty standardized though. I'm not really concerned about other LDAP servers.

Reply:
@smcintyre-r7 Hmm I think my opinion of LDAP may be different from yours as I'm finding there is a lot of parts of the documentation that leave it open to the servers how they wish to implement things. That being said, looking at your output above I would agree that having more filtering would be best. 

I'll see if I can't write something that does this filtering using a JSON file like you mentioned.

Reply:
Initial implementation now up at https://github.com/rapid7/metasploit-framework/pull/16598/commits/4e9f11724681ab50c3b3163f8e238f915d6ae325. Should be enough to prove concept, still need to code pulling stuff from a JSON file so we can do custom columns but this should prove the idea is viable and allow for a base set of queries that appropriately filter the data they return down to important bits.

Reply:
I assumed the output would be multiple tables, with 2 name/attribute columns, and multiple rows? i.e. similar to the output of ldapsearch

```
$ ldapsearch -H ldap://dc3.adf3.local -D "Administrator@adf3.local" -w 'p4$$w0rd1' -b 'dc=adf3,dc=local' '(|(objectClass=inetOrgPerson)(objectClass=user)(sAMAccountType=805306368)(objectClass=posixAccount)(objectClass=GsAccount)(objectClass=GsSIPUser))' '*'
# ...

# Administrator, Users, adf3.local
dn: CN=Administrator,CN=Users,DC=adf3,DC=local
objectClass: top
objectClass: person
objectClass: organizationalPerson
objectClass: user
cn: Administrator
description: Built-in account for administering the computer/domain
distinguishedName: CN=Administrator,CN=Users,DC=adf3,DC=local
instanceType: 4
whenCreated: 20220127114128.0Z
whenChanged: 20220624143534.0Z
uSNCreated: 8196
memberOf: CN=Group Policy Creator Owners,CN=Users,DC=adf3,DC=local
memberOf: CN=Domain Admins,CN=Users,DC=adf3,DC=local
memberOf: CN=Enterprise Admins,CN=Users,DC=adf3,DC=local
memberOf: CN=Schema Admins,CN=Users,DC=adf3,DC=local
memberOf: CN=Administrators,CN=Builtin,DC=adf3,DC=local
uSNChanged: 155817
name: Administrator
objectGUID:: O6huCET3fECrRABdDa5Egw==
userAccountControl: 512
badPwdCount: 0
codePage: 0
countryCode: 0
badPasswordTime: 133005907756586264
lastLogoff: 0
lastLogon: 133005907793466294
logonHours:: ////////////////////////////
pwdLastSet: 133004601821240246
primaryGroupID: 513
objectSid:: AQUAAAAAAAUVAAAA24l4SxXAM5DxwqRu9AEAAA==
adminCount: 1
accountExpires: 0
logonCount: 183
sAMAccountName: Administrator
sAMAccountType: 805306368
objectCategory: CN=Person,CN=Schema,CN=Configuration,DC=adf3,DC=local
isCriticalSystemObject: TRUE
dSCorePropagationData: 20220224104953.0Z
dSCorePropagationData: 20220127120215.0Z
dSCorePropagationData: 20220127120215.0Z
dSCorePropagationData: 20220127114705.0Z
dSCorePropagationData: 16010714042016.0Z
lastLogonTimestamp: 133004601821084478
msDS-SupportedEncryptionTypes: 0

# Guest, Users, adf3.local
dn: CN=Guest,CN=Users,DC=adf3,DC=local
objectClass: top
objectClass: person
objectClass: organizationalPerson
objectClass: user
cn: Guest
description: Built-in account for guest access to the computer/domain
distinguishedName: CN=Guest,CN=Users,DC=adf3,DC=local
instanceType: 4
whenCreated: 20220127114128.0Z
whenChanged: 20220127114128.0Z
uSNCreated: 8197
memberOf: CN=Guests,CN=Builtin,DC=adf3,DC=local
uSNChanged: 8197
name: Guest
objectGUID:: 6FmUXvNwY0OGIJSOZfUVQg==
userAccountControl: 66082
badPwdCount: 0
codePage: 0
countryCode: 0
badPasswordTime: 0
lastLogoff: 0
lastLogon: 0
pwdLastSet: 0
primaryGroupID: 514
objectSid:: AQUAAAAAAAUVAAAA24l4SxXAM5DxwqRu9QEAAA==
accountExpires: 9223372036854775807
logonCount: 0
sAMAccountName: Guest
sAMAccountType: 805306368
objectCategory: CN=Person,CN=Schema,CN=Configuration,DC=adf3,DC=local
isCriticalSystemObject: TRUE
dSCorePropagationData: 20220224104953.0Z
dSCorePropagationData: 20220127114705.0Z
dSCorePropagationData: 16010101000417.0Z

... etc ...
```

And I imagine folk would want the ability to output as different format types, i.e. as json, table, and potentially csv - similar to the Kubernetes enum module:

```
msf6 auxiliary(cloud/kubernetes/enum_kubernetes) > pod namespace=default name=redis-7fd956df5-sbchb output=json 
[*] Running module against 127.0.0.1

[
  {
    "kind": "Pod",
    "apiVersion": "v1",
    "metadata": {
      "name": "redis-7fd956df5-sbchb",
      "generateName": "redis-7fd956df5-",
      "namespace": "default",
      "uid": "0f00c08c-bdb1-4206-94ce-5c447cd2d446",
      "resourceVersion": "629723",
      "creationTimestamp": "2021-09-16T22:33:33Z",
      "labels": {
        "app": "redis",
        "pod-template-hash": "7fd956df5",
        "role": "leader",
        "tier": "backend"
      },
    },
    ... etc ...
  }
]
[*] Auxiliary module execution completed
```

Reply:
So for starters here, you and Spencer both seem to have different ideas on what the ideal output of this module should be. Whilst I think you both have valid points, I would like to see some mutual consensus here otherwise I'll be undoing and redoing things here and ultimately risking wasting developer time.

As to your question above though, we could certainly do the output in a way similar to `ldapsearch` if desired, however Spencer's point would still stand: if we did it this way it would just be multiple tables of 200+ entries. We would still need a way to filter things down.

I went with Rex::Table based off of Spencer's suggestion to do so however if we have an alternative way of doing this that you think would be dynamic enough to handle the various scenarios this module may face, then I'm all ears, however we should probably run this by the group if this will be influencing future LDAP work.

Reply:
> I went with Rex::Table based off of Spencer's suggestion to do

We'd still use Rex::Table, it would just be multiple tables 👀 

>  we could certainly do the output in a way similar to ldapsearch if desired, however Spencer's point would still stand: if we did it this way it would just be multiple tables of 200+ entries. We would still need a way to filter things down.

I don't think the example given conflicts with Spencer's statement. You would continue to default properties to the 3-4 useful values that make sense for the query you're providing out of the box, instead of `*` like in the ldapsearch example I gave. Then we'd provide an optional `properties=a,b,c` or `attrs=...` option which can override that. That would give folk the flexibility to specify `properties=*` if they want to.

> I would like to see some mutual consensus here

More than happy to jump on a call or talk about it during module hacking; Or if you want to take an executive decision that works for me too.

Reply:
Thanks guys will take a look over this and take some time to process everything, then will try come up with a proposed plan of action 👍

Reply:
> ```yaml
> (|(objectClass=inetOrgPerson)(objectClass=user)(sAMAccountType=805306368)(objectClass=posixAccount)(objectClass=GsAccount)(objectClass=GsSIPUser))
> ```

Added suggestion in with a24298c57dc03379d3d1c311194c93fda5d0b62a.

Reply:
The data file is close to what I had in mind, but I was thinking more that *all* of the queries would defined in it, and the action names would be loaded from the data file. This would allow users to modify the default queries we're shipping (to for example add additional filtering or change the columns) while also providing a template they could copy and paste from to add their own. Does that make sense?

When the module class is initialized, you'd load the data file and populate the actions there with the names / descriptions from the loaded data set.

Reply:
> I assumed the output would be multiple tables, with 2 columns (name and attribute), and multiple rows? i.e. similar to the output of ldapsearch
> 
> ```
> $ ldapsearch -H ldap://dc3.adf3.local -D "Administrator@adf3.local" -w 'p4$$w0rd1' -b 'dc=adf3,dc=local' '(|(objectClass=inetOrgPerson)(objectClass=user)(sAMAccountType=805306368)(objectClass=posixAccount)(objectClass=GsAccount)(objectClass=GsSIPUser))' '*'
> # ...
> 
> # Administrator, Users, adf3.local
> dn: CN=Administrator,CN=Users,DC=adf3,DC=local
> objectClass: top
> objectClass: person
> objectClass: organizationalPerson
> objectClass: user
> cn: Administrator
> description: Built-in account for administering the computer/domain
> distinguishedName: CN=Administrator,CN=Users,DC=adf3,DC=local
> instanceType: 4
> whenCreated: 20220127114128.0Z
> whenChanged: 20220624143534.0Z
> uSNCreated: 8196
> memberOf: CN=Group Policy Creator Owners,CN=Users,DC=adf3,DC=local
> memberOf: CN=Domain Admins,CN=Users,DC=adf3,DC=local
> memberOf: CN=Enterprise Admins,CN=Users,DC=adf3,DC=local
> memberOf: CN=Schema Admins,CN=Users,DC=adf3,DC=local
> memberOf: CN=Administrators,CN=Builtin,DC=adf3,DC=local
> uSNChanged: 155817
> name: Administrator
> objectGUID:: O6huCET3fECrRABdDa5Egw==
> userAccountControl: 512
> badPwdCount: 0
> codePage: 0
> countryCode: 0
> badPasswordTime: 133005907756586264
> lastLogoff: 0
> lastLogon: 133005907793466294
> logonHours:: ////////////////////////////
> pwdLastSet: 133004601821240246
> primaryGroupID: 513
> objectSid:: AQUAAAAAAAUVAAAA24l4SxXAM5DxwqRu9AEAAA==
> adminCount: 1
> accountExpires: 0
> logonCount: 183
> sAMAccountName: Administrator
> sAMAccountType: 805306368
> objectCategory: CN=Person,CN=Schema,CN=Configuration,DC=adf3,DC=local
> isCriticalSystemObject: TRUE
> dSCorePropagationData: 20220224104953.0Z
> dSCorePropagationData: 20220127120215.0Z
> dSCorePropagationData: 20220127120215.0Z
> dSCorePropagationData: 20220127114705.0Z
> dSCorePropagationData: 16010714042016.0Z
> lastLogonTimestamp: 133004601821084478
> msDS-SupportedEncryptionTypes: 0
> 
> # Guest, Users, adf3.local
> dn: CN=Guest,CN=Users,DC=adf3,DC=local
> objectClass: top
> objectClass: person
> objectClass: organizationalPerson
> objectClass: user
> cn: Guest
> description: Built-in account for guest access to the computer/domain
> distinguishedName: CN=Guest,CN=Users,DC=adf3,DC=local
> instanceType: 4
> whenCreated: 20220127114128.0Z
> whenChanged: 20220127114128.0Z
> uSNCreated: 8197
> memberOf: CN=Guests,CN=Builtin,DC=adf3,DC=local
> uSNChanged: 8197
> name: Guest
> objectGUID:: 6FmUXvNwY0OGIJSOZfUVQg==
> userAccountControl: 66082
> badPwdCount: 0
> codePage: 0
> countryCode: 0
> badPasswordTime: 0
> lastLogoff: 0
> lastLogon: 0
> pwdLastSet: 0
> primaryGroupID: 514
> objectSid:: AQUAAAAAAAUVAAAA24l4SxXAM5DxwqRu9QEAAA==
> accountExpires: 9223372036854775807
> logonCount: 0
> sAMAccountName: Guest
> sAMAccountType: 805306368
> objectCategory: CN=Person,CN=Schema,CN=Configuration,DC=adf3,DC=local
> isCriticalSystemObject: TRUE
> dSCorePropagationData: 20220224104953.0Z
> dSCorePropagationData: 20220127114705.0Z
> dSCorePropagationData: 16010101000417.0Z
> 
> ... etc ...
> ```
> 
> And I imagine folk would want the ability to output as different format types, i.e. as json, table, and potentially csv - similar to the Kubernetes enum module:
> 
> ```
> msf6 auxiliary(cloud/kubernetes/enum_kubernetes) > pod namespace=default name=redis-7fd956df5-sbchb output=json 
> [*] Running module against 127.0.0.1
> 
> [
>   {
>     "kind": "Pod",
>     "apiVersion": "v1",
>     "metadata": {
>       "name": "redis-7fd956df5-sbchb",
>       "generateName": "redis-7fd956df5-",
>       "namespace": "default",
>       "uid": "0f00c08c-bdb1-4206-94ce-5c447cd2d446",
>       "resourceVersion": "629723",
>       "creationTimestamp": "2021-09-16T22:33:33Z",
>       "labels": {
>         "app": "redis",
>         "pod-template-hash": "7fd956df5",
>         "role": "leader",
>         "tier": "backend"
>       },
>     },
>     ... etc ...
>   }
> ]
> [*] Auxiliary module execution completed
> ```

This should now be implemented with https://github.com/rapid7/metasploit-framework/pull/16598/commits/3ac6cd2fed4b0a54382e38f6fc71201c6b5ea63e and the CSV output should be implemented with https://github.com/rapid7/metasploit-framework/pull/16598/commits/34786c5f695be85a54356366cfbde00f398aa2ee. Thanks for the Slack chats that helped implement this @adfoster-r7!